docs/lab/05/index.md

```diff
@@ -0,0 +1,224 @@
+```rust
+loop {
+    info!("waiting for button press");
+    button.wait_for_falling_edge().await;
+}
+```
+We can see that here, we also use the `wait_for_falling_edge` function asynchronously, meaning that until we get a signal that a button has been pressed, the executor can decide to do other stuff. If it has nothing else to do, it goes to sleep, until it receives a signal that either the button has been pressed, or the timer has run out. Then, it will resume execution of the function where the action has completed. 
+When the button is pressed the execution flow will resume inside of the `button_pressed` task, until it is interrupted by the next `.await` in that function. If the timer runs out before the `button_pressed` task execution reaches the next `.await`, the resuming of the `main` function will delayed until the `button_pressed` task `.await`s.
+This method of development allows our programs to run seemingly "in parallel", without the need of multiple threads. Each task *voluntarily* pauses its execution and passes control over to whatever other task needs it. This means that it's the task's business to allow other tasks to run while it's idly waiting for something to happen on its end.
+:::note
+On an operating system, it would be the scheduler's job to decide when and for how long processes get to run.
+:::
+
+## Futures
+
+Rust has a special datatype that represents an action that will complete sometime in the future. By using `await` on a `Future`, we are passing control to another task until the Future completes.
+:::info
+In the `button_pressed` task, the `wait_for_falling_edge` returns a `Future`, which is then `await`ed.
+:::
+This is a simplified version of what a `Future` in Rust really looks like:
+```rust
+enum Poll<T> {
+    Pending,
+    Ready(T),
+}
+trait Future {
+   type Output;
+   fn poll(&mut self) -> Poll<Self::Output>;
+}
+```
+The `Future` has an `Output` type, that represents the type of the result that it will return once it completes. For `wait_for_falling_edge()`, the Output type is `()` (nothing).
+The function `poll` returns a Poll type, which can either be `Pending`, or `Ready<T>` (T will be output in this case).
+Let's break down what all of this means. A `Future` needs to be checked on, every now and then, to see what its status is. This is the job of the **Executor**. The executor must regularly ask the `Future` if it's completed, or if it needs more time before it can give a result. We can say that the `Future` is `poll`ed, and depending on whether it's ready to give a result or not, it gives its status as `Pending` or `Ready`. If it's still pending, it needs more time before it can return a result, so the executor moves on to poll another `Future`. Whenever the `Future` is completed, it returns `Ready` once polled, and the executor returns execution back to the function where the `Future` completed.
```

Comment:
```suggestion
Let's break down what all of this means. A `Future` needs to be checked on, every now and then, to see what its status is. This is the job of the **Executor**. The executor must regularly ask the `Future` if it's completed, or if it needs more time before it can give a result. We can say that the `Future` is `poll`ed, and depending on whether it's ready to give a result or not, it gives its status as `Pending` or `Ready`. If it's still pending, it needs more time before it can return a result, so the executor moves on to poll another `Future`. Whenever the `Future` is completed, it returns `Ready` once polled, and the executor returns execution back to the function where the `Future` `.await`ed.
```